File: messagebus/src/vespa/messagebus/network/rpcsendv2.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "rpcsendv2.h"
#include "rpcnetwork.h"
#include "rpcserviceaddress.h"
#include <vespa/messagebus/emptyreply.h>
#include <vespa/messagebus/tracelevel.h>
#include <vespa/vespalib/util/stringfmt.h>
#include <vespa/vespalib/data/slime/slime.h>
#include <vespa/vespalib/data/databuffer.h>
#include <vespa/vespalib/util/compressor.h>

using vespalib::make_string;
using vespalib::compression::CompressionConfig;
using vespalib::compression::decompress;
using vespalib::compression::compress;
using vespalib::DataBuffer;
using vespalib::ConstBufferRef;
using vespalib::stringref;
using vespalib::Memory;
using vespalib::Slime;
using vespalib::Version;
using namespace vespalib::slime;

namespace mbus {

namespace {

const char *METHOD_NAME   = "mbus.slime";
const char *METHOD_PARAMS = "bix";
const char *METHOD_RETURN = "sdISSsxs";

Memory VERSION_F("version");
Memory ROUTE_F("route");
Memory SESSION_F("session");
Memory USERETRY_F("use_retry");
Memory RETRY_F("retry");
Memory TIMELEFT_F("timeleft");
Memory PROTOCOL_F("prot");
Memory TRACELEVEL_F("trace");
Memory BLOB_F("msg");

}

bool RPCSendV2::isCompatible(stringref method, stringref request, stringref respons)
{
    return  (method == METHOD_NAME) &&
            (request == METHOD_PARAMS) &&
            (respons == METHOD_RETURN);
}

void
RPCSendV2::build(FRT_ReflectionBuilder & builder)
{
    builder.DefineMethod(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, true, FRT_METHOD(RPCSendV2::invoke), this);
    builder.MethodDesc("Send a message bus slime request and get a reply back.");
    builder.ParamDesc("encoding", "0=raw, 6=lz4");
    builder.ParamDesc("uncompressedBlobSize", "Uncompressed blob size");
    builder.ParamDesc("message", "The message blob in slime");
    builder.ReturnDesc("encoding",  "0=raw, 6=lz4");
    builder.ReturnDesc("uncompressedBlobSize", "Uncompressed blob size");
    builder.ReturnDesc("reply", "The reply blob in slime.");
}

const char *
RPCSendV2::getReturnSpec() const {
    return METHOD_RETURN;
}

namespace {
class OutputBuf : public vespalib::Output {
public:
    OutputBuf(size_t estimatedSize) : _buf(estimatedSize) { }
    DataBuffer & getBuf() { return _buf; }
private:
    vespalib::WritableMemory reserve(size_t bytes) override {
        _buf.ensureFree(bytes);
        return vespalib::WritableMemory(_buf.getFree(), _buf.getFreeLen());
    }
    Output &commit(size_t bytes) override {
        _buf.moveFreeToData(bytes);
        return *this;
    }
    DataBuffer _buf;
};
}

void
RPCSendV2::encodeRequest(FRT_RPCRequest &req, const Version &version, const Route & route,
                         const RPCServiceAddress & address, const Message & msg, uint32_t traceLevel,
                         const PayLoadFiller &filler, uint64_t timeRemaining) const
{
    FRT_Values &args = *req.GetParams();
    req.SetMethodName(METHOD_NAME);

    Slime slime;
    Cursor & root = slime.setObject();

    root.setString(VERSION_F, version.toString());
    root.setString(ROUTE_F, route.toString());
    root.setString(SESSION_F, address.getSessionName());
    root.setBool(USERETRY_F, msg.getRetryEnabled());
    root.setLong(RETRY_F, msg.getRetry());
    root.setLong(TIMELEFT_F, timeRemaining);
    root.setString(PROTOCOL_F, msg.getProtocol());
    root.setLong(TRACELEVEL_F, traceLevel);
    filler.fill(BLOB_F, root);

    OutputBuf buf(8192);
    BinaryFormat::encode(slime, buf);

    size_t unCompressedSize = buf.getBuf().getDataLen();
    args.AddInt8(0);
    args.AddInt32(buf.getBuf().getDataLen());
    args.AddData(buf.getBuf().stealBuffer(), unCompressedSize);
}

namespace {

class ParamsV2 : public RPCSend::Params
{
public:
    ParamsV2(const FRT_Values &arg)
        : _slime()
    {
        uint8_t encoding = arg[0]._intval8;
        uint32_t uncompressedSize = arg[1]._intval32;
        DataBuffer uncompressed(arg[2]._data._buf, arg[2]._data._len);
        ConstBufferRef blob(arg[2]._data._buf, arg[2]._data._len);
        decompress(CompressionConfig::toType(encoding), uncompressedSize, blob, uncompressed, true);
        assert(uncompressedSize == uncompressed.getDataLen());
        BinaryFormat::decode(Memory(uncompressed.getData(), uncompressed.getDataLen()), _slime);
    }

    uint32_t getTraceLevel() const override { return _slime.get()[TRACELEVEL_F].asLong(); }
    bool useRetry() const override { return _slime.get()[USERETRY_F].asBool(); }
    uint32_t getRetries() const override { return _slime.get()[RETRY_F].asLong(); }
    uint64_t getRemainingTime() const override { return _slime.get()[TIMELEFT_F].asLong(); }

    Version getVersion() const override {
        return Version(_slime.get()[VERSION_F].asString().make_stringref());
    }
    stringref getRoute() const override {
        return _slime.get()[ROUTE_F].asString().make_stringref();
    }
    stringref getSession() const override {
        return _slime.get()[SESSION_F].asString().make_stringref();
    }
    stringref getProtocol() const override {
        return _slime.get()[PROTOCOL_F].asString().make_stringref();
    }
    BlobRef getPayload() const override {
        Memory m = _slime.get()[BLOB_F].asData();
        return BlobRef(m.data, m.size);
    }
private:
    Slime _slime;
};

}

std::unique_ptr<RPCSend::Params>
RPCSendV2::toParams(const FRT_Values &args) const
{
    return std::make_unique<ParamsV2>(args);
}

std::unique_ptr<Reply>
RPCSendV2::createReply(const FRT_Values & ret, const string & serviceName,
                       Error & error, vespalib::TraceNode & rootTrace) const
{
    Version           version          = Version(ret[0]._string._str);
    double            retryDelay       = ret[1]._double;
    uint32_t         *errorCodes       = ret[2]._int32_array._pt;
    uint32_t          errorCodesLen    = ret[2]._int32_array._len;
    FRT_StringValue  *errorMessages    = ret[3]._string_array._pt;
    uint32_t          errorMessagesLen = ret[3]._string_array._len;
    FRT_StringValue  *errorServices    = ret[4]._string_array._pt;
    uint32_t          errorServicesLen = ret[4]._string_array._len;
    const char       *protocolName     = ret[5]._string._str;
    const char       *payload          = ret[6]._data._buf;
    uint32_t          payloadLen       = ret[6]._data._len;
    const char       *trace            = ret[7]._string._str;

    Reply::UP reply;
    if (payloadLen > 0) {
        reply = decode(protocolName, version, BlobRef(payload, payloadLen), error);
    }
    if ( ! reply ) {
        reply.reset(new EmptyReply());
    }
    reply->setRetryDelay(retryDelay);
    for (uint32_t i = 0; i < errorCodesLen && i < errorMessagesLen && i < errorServicesLen; ++i) {
        reply->addError(Error(errorCodes[i], errorMessages[i]._str,
                              errorServices[i]._len > 0 ? errorServices[i]._str : serviceName.c_str()));
    }
    rootTrace.addChild(TraceNode::decode(trace));
    return reply;
}

void
RPCSendV2::createResponse(FRT_Values & ret, const string & version, Reply & reply, Blob payload) const
{
    ret.AddString(version.c_str());
    ret.AddDouble(reply.getRetryDelay());

    uint32_t         errorCount    = reply.getNumErrors();
    uint32_t        *errorCodes    = ret.AddInt32Array(errorCount);
    FRT_StringValue *errorMessages = ret.AddStringArray(errorCount);
    FRT_StringValue *errorServices = ret.AddStringArray(errorCount);
    for (uint32_t i = 0; i < errorCount; ++i) {
        errorCodes[i] = reply.getError(i).getCode();
        ret.SetString(errorMessages + i, reply.getError(i).getMessage().c_str());
        ret.SetString(errorServices + i, reply.getError(i).getService().c_str());
    }

    ret.AddString(reply.getProtocol().c_str());
    ret.AddData(std::move(payload.payload()), payload.size());
    if (reply.getTrace().getLevel() > 0) {
        ret.AddString(reply.getTrace().getRoot().encode().c_str());
    } else {
        ret.AddString("");
    }
}

} // namespace mbus

File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCSendV2.java
package com.yahoo.messagebus.network.rpc;

import com.yahoo.component.Version;
import com.yahoo.compress.CompressionType;
import com.yahoo.compress.Compressor;
import com.yahoo.jrt.DataValue;
import com.yahoo.jrt.DoubleValue;
import com.yahoo.jrt.Int32Array;
import com.yahoo.jrt.Int32Value;
import com.yahoo.jrt.Int64Value;
import com.yahoo.jrt.Int8Value;
import com.yahoo.jrt.Method;
import com.yahoo.jrt.Request;
import com.yahoo.jrt.StringArray;
import com.yahoo.jrt.StringValue;
import com.yahoo.jrt.Values;
import com.yahoo.messagebus.EmptyReply;
import com.yahoo.messagebus.Error;
import com.yahoo.messagebus.Message;
import com.yahoo.messagebus.Reply;
import com.yahoo.messagebus.Trace;
import com.yahoo.messagebus.TraceNode;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.slime.BinaryFormat;
import com.yahoo.slime.Cursor;
import com.yahoo.slime.Inspector;
import com.yahoo.slime.Slime;
import com.yahoo.text.Utf8;
import com.yahoo.text.Utf8Array;

/**
 * Implements the request adapter for method "mbus.send1".
 *
 * @author <a href="mailto:simon@yahoo-inc.com">Simon Thoresen</a>
 */
public class RPCSendV2 extends RPCSend {

    private final static String METHOD_NAME = "mbus.slime";
    private final static String METHOD_PARAMS = "bix";
    private final static String METHOD_RETURN = "sdISSsxs";
    private final Compressor compressor = new Compressor(CompressionType.LZ4, 3, 90, 1024);

    @Override
    protected String getReturnSpec() { return METHOD_RETURN; }
    @Override
    protected Method buildMethod() {

        Method method = new Method(METHOD_NAME, METHOD_PARAMS, METHOD_RETURN, this);
        method.methodDesc("Send a message bus request and get a reply back.");
        method.paramDesc(0, "encoding", "Encoding type.")
                .paramDesc(1, "decodedSize", "Number of bytes after decoding.")
                .paramDesc(2, "payload", "Slime encoded payload.");
        method.returnDesc(0, "version", "The lowest version the message was serialized as.")
                .returnDesc(1, "retryDelay", "The retry request of the reply.")
                .returnDesc(2, "errorCodes", "The reply error codes.")
                .returnDesc(3, "errorMessages", "The reply error messages.")
                .returnDesc(4, "errorServices", "The reply error service names.")
                .returnDesc(5, "protocol", "The name of the protocol that knows how to decode this reply.")
                .returnDesc(6, "payload", "The protocol specific reply payload.")
                .returnDesc(7, "trace", "A string representation of the trace.");
        return method;
    }
    private static final String VERSION_F = new String("version");
    private static final String ROUTE_F = new String("route");
    private static final String SESSION_F = new String("session");
    private static final String PROTOCOL_F = new String("prot");
    private static final String TRACELEVEL_F = new String("tracelevel");
    private static final String USERETRY_F = new String("useretry");
    private static final String RETRY_F = new String("retry");
    private static final String TIMEREMAINING_F = new String("timeleft");
    private static final String BLOB_F = new String("msg");



    @Override
    protected Request encodeRequest(Version version, Route route, RPCServiceAddress address, Message msg,
                                    long timeRemaining, byte[] payload, int traceLevel) {
        Slime slime = new Slime();
        Cursor root = slime.setObject();

        root.setString(VERSION_F, version.toString());
        root.setString(ROUTE_F, route.toString());
        root.setString(SESSION_F, address.getSessionName());
        root.setString(PROTOCOL_F, msg.getProtocol().toString());
        root.setBool(USERETRY_F, msg.getRetryEnabled());
        root.setLong(RETRY_F, msg.getRetry());
        root.setLong(TIMEREMAINING_F, msg.getTimeRemaining());
        root.setLong(TRACELEVEL_F, traceLevel);
        root.setData(BLOB_F, payload);

        byte[] serializedSlime = BinaryFormat.encode(slime);
        Compressor.Compression compressionResult = compressor.compress(serializedSlime);
        Request req = new Request(METHOD_NAME);
        Values v = req.parameters();

        v.add(new Int8Value(compressionResult.type().getCode()));
        v.add(new Int32Value(compressionResult.uncompressedSize()));
        v.add(new DataValue(compressionResult.data()));

        return req;
    }

    @Override
    protected Reply createReply(Request req, String serviceName, Trace trace) {
        // Retrieve all reply components from JRT request object.
        Version version = new Version(req.returnValues().get(0).asUtf8Array());
        double retryDelay = req.returnValues().get(1).asDouble();
        int[] errorCodes = req.returnValues().get(2).asInt32Array();
        String[] errorMessages = req.returnValues().get(3).asStringArray();
        String[] errorServices = req.returnValues().get(4).asStringArray();
        Utf8Array protocolName = req.returnValues().get(5).asUtf8Array();
        byte[] payload = req.returnValues().get(6).asData();
        String replyTrace = req.returnValues().get(7).asString();

        // Make sure that the owner understands the protocol.
        Reply reply = null;
        Error error = null;
        if (payload.length > 0) {
            Object ret = decode(protocolName, version, payload);
            if (ret instanceof Reply) {
                reply = (Reply) ret;
            } else {
                error = (Error) ret;
            }
        }
        if (reply == null) {
            reply = new EmptyReply();
        }
        if (error != null) {
            reply.addError(error);
        }
        reply.setRetryDelay(retryDelay);
        for (int i = 0; i < errorCodes.length && i < errorMessages.length; i++) {
            reply.addError(new Error(errorCodes[i], errorMessages[i],
                    errorServices[i].length() > 0 ? errorServices[i] : serviceName));
        }
        if (trace.getLevel() > 0) {
            trace.getRoot().addChild(TraceNode.decode(replyTrace));
        }
        return reply;
    }

    protected Params toParams(Values args) {
        CompressionType compression = CompressionType.valueOf(args.get(0).asInt8());
        byte[] slimeBytes = compressor.decompress(args.get(2).asData(), compression, args.get(1).asInt32());
        Slime slime = BinaryFormat.decode(slimeBytes);
        Inspector root = slime.get();
        Params p = new Params();
        p.version = new Version(root.field(VERSION_F).asString());
        p.route = root.field(ROUTE_F).asString();
        p.session = root.field(SESSION_F).asString();
        p.retryEnabled = root.field(USERETRY_F).asBool();
        p.retry = (int)root.field(RETRY_F).asLong();
        p.timeRemaining = root.field(TIMEREMAINING_F).asLong();
        p.protocolName = new Utf8Array(Utf8.toBytes(root.field(PROTOCOL_F).asString()));
        p.payload = root.field(BLOB_F).asData();
        p.traceLevel = (int)root.field(TRACELEVEL_F).asLong();
        return p;
    }

    @Override
    protected void createReponse(Values ret, Reply reply, Version version, byte [] payload) {
        int[] eCodes = new int[reply.getNumErrors()];
        String[] eMessages = new String[reply.getNumErrors()];
        String[] eServices = new String[reply.getNumErrors()];
        for (int i = 0; i < reply.getNumErrors(); ++i) {
            Error error = reply.getError(i);
            eCodes[i] = error.getCode();
            eMessages[i] = error.getMessage();
            eServices[i] = error.getService() != null ? error.getService() : "";
        }
        ret.add(new StringValue(version.toString()));
        ret.add(new DoubleValue(reply.getRetryDelay()));
        ret.add(new Int32Array(eCodes));
        ret.add(new StringArray(eMessages));
        ret.add(new StringArray(eServices));
        ret.add(new StringValue(reply.getProtocol()));
        ret.add(new DataValue(payload));
        ret.add(new StringValue(reply.getTrace().getRoot() != null ? reply.getTrace().getRoot().encode() : ""));
    }

}
